### ISSUE TITLE
WoodFisher: Efficient Second-Order Approximation for Neural Network Compression

### ISSUE THREAD

**Issue by Hiroki11x** [Hiroki11x/Papers#64]
## NeurIPS2020気になったシリーズ

https://papers.nips.cc/paper/2020/hash/d1ff1ec86b62cd5f3903ff19c3a326b2-Abstract.html

> ヘシアンまたは逆ヘシアン-ベクトル積の形をした2次情報は、最適化問題を解くための基本的なツールです。近年、この情報をディープニューラルネットワークで利用することが注目されていますが、既存の近似の品質についてはほとんど知られていません。</p> <p>主な用途はニューラル ネットワークの圧縮で、古典的な Optimal Brain Damage/Surgeon フレームワークをベースにしています。これにより、WoodFisherがワンショット剪定のための最先端の一般的な手法を大幅に凌駕することを実証しています。さらに、反復的で段階的な剪定が可能な場合でも、ImageNet ILSVRCのような一般的な画像分類データに対しては、最先端の手法と比較してテスト精度を向上させることができた。さらに，我々の手法が一次情報を考慮に入れるためにどのように拡張できるかを示し，レイヤごとの剪定しきい値を自動的に設定したり，限られたデータ領域で圧縮を行ったりする能力を示す．

<img width="753" alt="Screen Shot 2020-12-04 at 16 46 50" src="https://user-images.githubusercontent.com/8721858/101136203-52d66d00-3650-11eb-87de-fb592553b2a9.png">

<img width="771" alt="Screen Shot 2020-12-04 at 16 43 05" src="https://user-images.githubusercontent.com/8721858/101135931-e3f91400-364f-11eb-8a98-af5935a38f13.png">
 
# 最近思ってた疑問
![128599919_423382115479800_6674521360639297141_n](https://user-images.githubusercontent.com/8721858/101137847-c8dbd380-3652-11eb-8126-569f6f14c19b.jpg)
![129134491_191590069238493_2113568229764806156_n](https://user-images.githubusercontent.com/8721858/101137853-caa59700-3652-11eb-8b5d-2ccb2b5ed38d.jpg)

つまり FisherをP_{y,x}で期待値取りたいけど
P_{y,x} = Q_{x}P_{y|x}　みたいになって、
 
Q_{x}は知り得ないので、\hat{Q_{x}}として、

かつ、P_{y|x}が、One Hot Vectorの答えの、\hat{Q_{y|x}} に一致するとしたら(これは十分KLが近いと仮定すれば)
FとHは一致する、それはそう。それって、　Empirical Fisherのことではないのか？？！

<img width="583" alt="Screen Shot 2020-12-04 at 17 11 42" src="https://user-images.githubusercontent.com/8721858/101138812-05f49580-3654-11eb-8429-c7f48cbbfdb7.png">
https://www.inference.vc/on-empirical-fisher-information/

そこで下記を読んで見る
Limitations of the Empirical Fisher Approximation for Natural Gradient Descent
https://arxiv.org/abs/1905.12558

<img width="707" alt="Screen Shot 2020-12-04 at 17 17 05" src="https://user-images.githubusercontent.com/8721858/101139457-ead65580-3654-11eb-9c0f-81ef367ddbb3.png">

MLでは、データ方向の期待値に関してはしょうがないと言った印象
<img width="728" alt="Screen Shot 2020-12-04 at 17 20 18" src="https://user-images.githubusercontent.com/8721858/101139545-093c5100-3655-11eb-9768-3b267f4e886f.png">

ここで、　Empirical Fisherは (y_nとなってるのはラベルであることを示すため)

<img width="721" alt="Screen Shot 2020-12-04 at 17 22 58" src="https://user-images.githubusercontent.com/8721858/101139862-810a7b80-3655-11eb-9960-3552abb9c949.png">

経験的フィッシャー(Eq. 7)がフィッシャー(Eq. 6)にどれだけ近いかは、モデルpθ(y|xn)が真のデータ生成分布p(y|xn)にどれだけ近いかに依存します。

#### Conclusion

> EFは一般化ガウス・ニュートン行列の正式な定義に従っているが、基礎となる分割は有用な二次情報を保持していない。我々はGGNの定義を明確に修正することを提案した。

> 経験的フィッシャーとフィッシャーの間の明確な関係は，強い追加の仮定の下でのみ最低でも存在する．(i) 正しいモデルと(ii) モデルのキャパシティ。

> これらの条件は、実際には満たされる可能性は低く、特に オーバーパラメトリック化された一般関数近似器と近似最小値のための設定。

## Efficient Estimates of Inverse-Hessian Vector Products

<img width="764" alt="Screen Shot 2020-12-04 at 16 43 40" src="https://user-images.githubusercontent.com/8721858/101136112-2f132700-3650-11eb-904f-83aaa1d79681.png">
